### PR TITLE
[Pass] Simplify Batch Norm in Training Mode

### PR DESCRIPTION
**Pull request by SiriusNEO**
**Background**

Prior to this PR, we have the pass `SimplifyNormInference` which helps us unpack the batch norm computation in index 0. But since the gradient of BN is complex and a good solution is to simplify it to smaller operators, it is necessary that we should introduce simplify for training, not just for inference.

**Work**

This PR changes the pass `SimplifyNormInference` to `SimplifyNorm`, and add an argument `mode: Literal["eval", "training"]` to indicate the simplification mode.

### CODE REVIEW

File: python/tvm/relax/transform/transform.py [hunk 1]
```diff
@@ -530,18 +534,26 @@ def MetaScheduleTuneIRMod(
...
+
+    During inference, for example, the result of a batch norm which is indexed at
+    tuple index 0 will be unpacked into a number of simplified operators.
+
+    During training, the result of a batch norm in all indices will be unpacked.
+
+    Parameters
+    ----------
+    mode: Literal["eval", "training"]
+        The mode of simplification. Can be `eval` or `training`.
```
Comment (Ubospica) [inline python/tvm/relax/transform/transform.py]:
```suggestion
        The mode of simplification. Can be `"eval"` or `"training"`.
```

File: python/tvm/relax/transform/transform.py [hunk 2]
```diff
@@ -916,7 +928,7 @@ def dataflowblock_pass(
         name of the optimization function will be used as the pass name.
 
     required : Optional[List[str]]
-        The list of passes that the dataflowblock pass is dependent on.
+        The list of passes that the dataflowblock pass is dependent on.z
```
Comment (Ubospica) [inline python/tvm/relax/transform/transform.py]:
```suggestion
        The list of passes that the dataflowblock pass is dependent on.
```

File: src/relax/transform/simplify_norm.cc [hunk 3]
```diff
@@ -73,10 +74,60 @@ Expr SimplifyBatchNorm(const CallNode* call) {
...
+}
+
+static const char* kModeEval = "eval";
+static const char* kModeTraining = "training";
+
+/*! \brief A mutator to simplify the normalization. */
+class NormSimplifier : public ExprMutator {
  public:
-  static Expr Simplify(Expr expr) { return NormInferenceSimplifier()(expr); }
+  static Expr Simplify(Expr expr, String mode) { return NormSimplifier(mode)(expr); }
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: You can check mode is in ['eval', 'training'] and report error in Simplify()

Cuz checking it in wrapper function is better than in `VisitBinding_` I think

File: src/relax/transform/simplify_norm.cc [hunk 4]
```diff
@@ -101,30 +152,37 @@ class NormInferenceSimplifier : public ExprMutator {
       // Instead, we will store the unpacked value in the batch_norm_map_, and replace it
       // at the TupleGetItemNode. And the original batch_norm call will be removed in the
       // follow-up pass `RemoveAllUnused`
-      batch_norm_map_.Set(binding->var, SimplifyBatchNorm(val));
+      if (mode_ == kModeEval) {
+        batch_norm_map_.Set(binding->var, {SimplifyBatchNormInference(val), NullOpt, NullOpt});
+      } else {
+        ICHECK_EQ(mode_, kModeTraining);
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc:163]: As above, you can keep this line or just remove it

File: src/relax/transform/simplify_norm.cc [hunk 5]
```diff
@@ -73,10 +74,60 @@ Expr SimplifyBatchNorm(const CallNode* call) {
...
+    }
+  }
+
+  Expr data_mean = mean(data, reduce_axes, false);
+  Expr data_mean_rs = ExpandToMatchInput(data_mean, sinfo->ndim, {attrs->axis});
+  Expr data_var = variance(data, reduce_axes, false);
+  Expr data_var_rs = ExpandToMatchInput(data_var, sinfo->ndim, {attrs->axis});
+
+  // output = (x - mean) / sqrt(var + epsilon) * gamma + beta
+  Expr epsilon = MakeConstantScalar(static_cast<float>(attrs->epsilon), sinfo->dtype);
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: No need to static_cast here

File: src/relax/transform/simplify_norm.cc [hunk 6]
```diff
@@ -73,10 +74,60 @@ Expr SimplifyBatchNorm(const CallNode* call) {
   return out;
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: Maybe we can simplify running_mean and var as well?

Then 
1. We do not need `Array<Optional<Expr>>`, just `Array<Expr>` as return type
2. We can directly remove the batchnorm binding, not have to RemoveAllUnused
3. no batchnorm will exist after transform

File: src/relax/transform/simplify_norm.cc [hunk 7]
```diff
@@ -73,10 +74,60 @@ Expr SimplifyBatchNorm(const CallNode* call) {
...
+      add(multiply(one_minus_mom, moving_mean), multiply(momentum, data_mean)),
+      add(multiply(one_minus_mom, moving_var), multiply(momentum, data_var)),
+  };
+}
+
+static const char* kModeEval = "eval";
+static const char* kModeTraining = "training";
+
+/*! \brief A mutator to simplify the normalization. */
+class NormSimplifier : public ExprMutator {
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: We can privately inherit it

File: src/relax/transform/simplify_norm.cc [hunk 8]
```diff
@@ -73,10 +74,60 @@ Expr SimplifyBatchNorm(const CallNode* call) {
...
  public:
-  static Expr Simplify(Expr expr) { return NormInferenceSimplifier()(expr); }
+  static Expr Simplify(Expr expr, String mode) { return NormSimplifier(mode)(expr); }
+
+  explicit NormSimplifier(String mode) : ExprMutator(), mode_(mode) {}
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: This could be private method

Cuz the only interface is Simplify()

File: src/relax/transform/simplify_norm.cc [hunk 9]
```diff
@@ -101,30 +152,37 @@ class NormInferenceSimplifier : public ExprMutator {
...
+      }
     }
   }
 
  private:
   /*! \brief The mapping from binding var of batch_norm to the unpacked value. */
-  Map<Expr, Expr> batch_norm_map_;
+  Map<Expr, Array<Optional<Expr>>> batch_norm_map_;
+
+  String mode_;
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: could be const

File: src/relax/transform/simplify_norm.cc [hunk 10]
```diff
@@ -73,10 +74,60 @@ Expr SimplifyBatchNorm(const CallNode* call) {
...
+  if (attrs->scale) {
+    out = multiply(out, ExpandToMatchInput(gamma, sinfo->ndim, {attrs->axis}));
+  }
+  if (attrs->center) {
+    out = add(out, ExpandToMatchInput(beta, sinfo->ndim, {attrs->axis}));
+  }
+
+  Expr moving_mean = call->args[3];
+  Expr moving_var = call->args[4];
+  Expr momentum = MakeConstantScalar(static_cast<float>(attrs->momentum), sinfo->dtype);
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: Ditto

File: tests/python/relax/test_transform_simplify_norm.py [hunk 11]
```diff
@@ -146,7 +146,82 @@ def expected(
...
+        R.Tensor((64,), dtype="float32"),
+    ):
+        with R.dataflow():
+            lv: R.Tensor((64,), dtype="float32") = R.mean(x, axis=[0, 2, 3], keepdims=False)
+            lv1: R.Tensor((1, 64, 1, 1), dtype="float32") = R.expand_dims(lv, axis=[0, 2, 3])
+            lv2: R.Tensor((1, 64, 112, 112), dtype="float32") = R.subtract(x, lv1)
+            lv3: R.Tensor((64,), dtype="float32") = R.variance(x, axis=[0, 2, 3], keepdims=False)
+            lv4: R.Tensor((1, 64, 1, 1), dtype="float32") = R.expand_dims(lv3, axis=[0, 2, 3])
+            lv5: R.Tensor((1, 64, 1, 1), dtype="float32") = R.add(
+                lv4, R.const(9.9999997473787516e-06, "float32")
```
Comment (Ubospica) [inline tests/python/relax/test_transform_simplify_norm.py]: You can write the constants simpler, like 1e-5, 0.9, etc. That will make script easier to read, and assert_structural will not complain error

File: src/relax/transform/simplify_norm.cc [hunk 12]
```diff
@@ -73,10 +74,60 @@ Expr SimplifyBatchNorm(const CallNode* call) {
...
+  Expr one_minus_mom = MakeConstantScalar(static_cast<float>(1 - attrs->momentum), sinfo->dtype);
+
+  return {
+      out,
+      add(multiply(one_minus_mom, moving_mean), multiply(momentum, data_mean)),
+      add(multiply(one_minus_mom, moving_var), multiply(momentum, data_var)),
+  };
+}
+
+static const char* kModeEval = "eval";
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: We can define it as static constexpr const char*. That implies kModeEval pointer is constexpr

File: include/tvm/relax/transform.h [hunk 13]
```diff
@@ -273,12 +273,14 @@ TVM_DLL Pass RunCodegen(Optional<Map<String, Map<String, ObjectRef>>> target_opt
...
 
 /*!
- * \brief Simplify normalization operators during inference. For example, the result
- * of a batch norm which is indexed at tuple index 0 will be unpacked into a
- * number of simplified operators.
+ * \brief Simplify normalization operators.
+ * During inference, for example, the result of a batch norm which is indexed at
+ * tuple index 0 will be unpacked into a number of simplified operators.
+ * During training, the result of a batch norm in all indices will be unpacked.
+ * \param mode The mode of simplification. Can be `eval` or `training`.
```
Comment (Ubospica) [inline include/tvm/relax/transform.h:280]: Ditto

File: src/relax/transform/simplify_norm.cc (hunk 12 above)
Comment (Ubospica) [reply src/relax/transform/simplify_norm.cc]: And we can define it as public member of NormSimplifier because it is only used there

File: src/relax/transform/simplify_norm.cc [hunk 14]
```diff
@@ -101,30 +152,37 @@ class NormInferenceSimplifier : public ExprMutator {
...
       [=](Function f, IRModule m, PassContext pc) {
-        f = Downcast<Function>(NormInferenceSimplifier::Simplify(f));
+        f = Downcast<Function>(NormSimplifier::Simplify(f, mode));
         // Remove original batch_norm op if it's not used.
         return RemoveAllUnused(f);
       };
-  return CreateFunctionPass(/*pass_function=*/pass_func,            //
-                            /*opt_level=*/0,                        //
-                            /*pass_name=*/"SimplifyNormInference",  //
+  return CreateFunctionPass(/*pass_function=*/pass_func,   //
```
Comment (Ubospica) [inline src/relax/transform/simplify_norm.cc]: Sometimes we only need to modify one function in a irmodule. Shall we change it into ModulePass and pass a Optional<String> parameter that specifies the function to simplify, and if it's null, we will simplify all functions in the given IRModule?

**Review by Ubospica — APPROVED** [review]
